apps/mark-scan/backend/src/custom-paper-handler/application_driver.ts

```diff
@@ -0,0 +1,225 @@
+  );
+}
+
+export function isPaperInOutput(
+  paperHandlerStatus: PaperHandlerStatus
+): boolean {
+  return paperHandlerStatus.ticketPresentInOutput;
+}
+
+export function isPaperAnywhere(
+  paperHandlerStatus: PaperHandlerStatus
+): boolean {
+  return (
+    isPaperInInput(paperHandlerStatus) || isPaperInScanner(paperHandlerStatus)
+  );
+}
+
+export async function logRawStatus(driver: PaperHandlerDriver): Promise<void> {
+  debug('%O', await driver.getPaperHandlerStatus());
+}
+
+export async function setDefaults(driver: PaperHandlerDriver): Promise<void> {
+  await driver.initializePrinter();
+  debug('initialized printer (0x1B 0x40)');
+  await driver.setLineSpacing(0);
+  debug('set line spacing to 0');
+  await driver.setPrintingSpeed('slow');
+  debug('set printing speed to slow');
+}
+
+export async function printBallot(
+  driver: PaperHandlerDriver,
+  pdfData: Uint8Array,
+  options: Partial<ImageConversionOptions> = {}
+): Promise<void> {
+  const enablePrintPromise = driver.enablePrint();
+
+  let time = Date.now();
+  const pages: ImageData[] = [];
+  for await (const { page, pageCount } of pdfToImages(Buffer.from(pdfData), {
+    scale: 200 / 72,
+  })) {
+    assert(pageCount === 1, `Unexpected page count ${pageCount}`);
+    pages.push(page);
+  }
+  const page = pages[0];
+  assert(page, 'Unexpected undefined page');
+  debug(`pdf to image took ${Date.now() - time} ms`);
+  time = Date.now();
+
+  const ballotBinaryBitmap = imageDataToBinaryBitmap(page, options);
+  debug(`bitmap width: ${ballotBinaryBitmap.width}`);
+  debug(`bitmap height: ${ballotBinaryBitmap.height}`);
+  debug(`image to binary took ${Date.now() - time} ms`);
+  time = Date.now();
+
+  const customChunkedBitmaps = chunkBinaryBitmap(ballotBinaryBitmap);
+  debug(`num chunk rows: ${customChunkedBitmaps.length}`);
+  debug(`binary to chunks took ${Date.now() - time} ms`);
+
+  await enablePrintPromise;
+  debug('Begin printing');
+  let dotsSkipped = 0;
+  for (const customChunkedBitmap of customChunkedBitmaps) {
+    if (customChunkedBitmap.empty) {
+      dotsSkipped += VERTICAL_DOTS_IN_CHUNK;
+    } else {
+      if (dotsSkipped) {
+        await driver.setRelativeVerticalPrintPosition(dotsSkipped * 2); // assuming default vertical units, 1 / 408 units
+        dotsSkipped = 0;
+      }
+      await driver.printChunk(customChunkedBitmap);
+    }
+  }
+  debug('Completed printing %d chunks total', customChunkedBitmaps.length);
+}
+
+export async function scanAndSave(
+  driver: PaperHandlerDriver
+): Promise<SheetOf<string>> {
+  debug('+scanAndSave');
+  const dir = '/tmp';
+  const dateString = new Date().toISOString();
+  const pathOutFront = join(dir, `ballot-statemachine-${dateString}-A.jpg`);
+  // We can only print to one side from the thermal printer, but the interpret flow expects
+  // a SheetOf 2 pages. Use an image of a blank sheet for the 2nd page.
+  const blankSheetFixturePath = join(__dirname, 'fixtures', 'blank-sheet.jpg');
```

Comment:
It's unfortunate that we're still passing paths around instead of dealing with image data directly. If we were using image data directly we could just do `createImageData(width, height)` to make a blank page.

Reply:
To clarify, this would be blocked by a future refactor of the interpret pipeline ie. no action needed in this PR?